command/eval_delete.go

```diff
@@ -0,0 +1,529 @@
+		if err != nil {
+			return nil
+		}
+
+		resp, _, err := client.Search().PrefixSearch(a.Last, contexts.Evals, nil)
+		if err != nil {
+			return []string{}
+		}
+		return resp.Matches[contexts.Evals]
+	})
+}
+
+func (e *EvalDeleteCommand) Name() string { return "eval delete" }
+
+func (e *EvalDeleteCommand) Run(args []string) int {
+
+	flags := e.Meta.FlagSet(e.Name(), FlagSetClient)
+	flags.Usage = func() { e.Ui.Output(e.Help()) }
+	flags.Var(&e.scheduler, "scheduler", "")
+	flags.Var(&e.status, "status", "")
+	flags.Var(&e.job, "job", "")
+	flags.Var(&e.deployment, "deployment", "")
+	flags.Var(&e.node, "node", "")
+	flags.StringVar(&e.filterOp, "filter-operator", "and", "")
+	flags.BoolVar(&e.yes, "yes", false, "")
+	if err := flags.Parse(args); err != nil {
+		return 1
+	}
+
+	args = flags.Args()
+
+	if err := e.verifyArgsAndFlags(args); err != nil {
+		e.Ui.Error(fmt.Sprintf("Error validating command args and flags: %v", err))
+		return 1
+	}
+
+	// Pad the filter operator with whitespace, so it can be used as is when
+	// building the filter. This transforms "and" to " and ".
+	originalOP := e.filterOp
+	e.filterOp = " " + originalOP + " "
+
+	// Get the HTTP client and store this for use across multiple functions.
+	client, err := e.Meta.Client()
+	if err != nil {
+		e.Ui.Error(fmt.Sprintf("Error initializing client: %s", err))
+		return 1
+	}
+	e.client = client
+
+	// Pause the eval broker to ensure no more work can be dequeued and
+	// processed by the schedulers.
+	if err := e.pauseEvalBroker(); err != nil {
+		e.Ui.Error(fmt.Sprintf("Error pausing eval broker: %s", err))
+		return 1
+	}
+
+	// Track the eventual exit code as there are a number of factors that
+	// influence this.
+	var exitCode int
+
+	// Call the correct function in order to handle to operator input
+	// correctly.
+	switch len(args) {
+	case 1:
+		e.deleteByArg = true
+		exitCode, err = e.handleEvalArgDelete(args[0])
+	default:
+		exitCode, err = e.handleFlagFilterDelete("", e.buildFilter())
+	}
+
+	// Do not exit if we got an error as it's possible this was on the
+	// non-first iteration, and we have therefore deleted some evals.
+	if err != nil {
+		e.Ui.Error(fmt.Sprintf("Error deleting evaluations: %s", err))
+	}
+
+	// Depending on whether we deleted evaluations or not, output a message so
+	// this is clear.
+	if e.numDeleted > 0 {
+		grammar := "evaluation"
+		if e.numDeleted > 1 {
+			grammar = grammar + "s"
+		}
+		e.Ui.Output(fmt.Sprintf("Successfully deleted %v %s", e.numDeleted, grammar))
+	} else if err == nil {
+		e.Ui.Output("No evaluations were deleted")
+	}
+
+	// Always attempt to revert the state of the eval broker. In the event this
+	// succeeds, but the eval deletion failed, maintain the non-zero exit code
+	// so failures are clear.
+	return func(exitCode int) int {
+		if err := e.revertEvalBroker(); err != nil {
+			e.Ui.Error(fmt.Sprintf("Error reverting eval broker state: %s", err))
+			exitCode = 1
+		}
+		return exitCode
+	}(exitCode)
```

Comment:
I'll add some additional logging around the changing of the scheduler config state to help this problem. The alternative is to remove this state change from the command altogether, and require operators to do this before running this command. I don't feel this is the correct option though, as it adds additional overhead to operators when it's likely this will be used in stressful situations.

It's also worth noting that by default, the command when using filters go through two confirmation steps before each batch of deletions. This should help reduce the regret it moments caused.

Reply:
This has now been resolved within https://github.com/hashicorp/nomad/pull/13492/commits/432226048f5d0e83c357a68408e51da726502068